common/transport/amqp/src/receiver_link.ts

```diff
@@ -0,0 +1,212 @@
+
+            if (callback) {
+              callback(err);
+            } else if (err) {
+              /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_011: [If a `detached` or `errorReceived` event is emitted by the `amqp10` link object, the `ReceiverLink` object shall forward that error to the client.]*/
+              this.emit('error', err);
+            }
+          },
+          attach: (callback) => {
+            this._fsm.transition('attaching', callback);
+          },
+          detach: (callback) => { this._safeCallback(callback); },
+          forceDetach: () => { return; },
+          accept: (message, callback) => callback(new errors.DeviceMessageLockLostError()),
+          reject: (message, callback) => callback(new errors.DeviceMessageLockLostError()),
+          abandon: (message, callback) => callback(new errors.DeviceMessageLockLostError())
+        },
+        attaching: {
+          _onEnter: (callback) => {
+            /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_004: [The `attach` method shall use the stored instance of the `amqp10.AmqpClient` object to attach a new link object with the `linkAddress` and `linkOptions` provided when creating the `ReceiverLink` instance.]*/
+            debug('creating receiver with amqp10 for: ' + this._linkAddress);
+            this._amqp10Client.createReceiver(this._linkAddress, this._linkOptions)
+              .then((amqp10link) => {
+                debug('receiver object created by amqp10 for endpoint: ' + this._linkAddress);
+                if (this._fsm.state === 'attaching') {
+                  this._linkObject = amqp10link;
+                  /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_006: [The `ReceiverLink` object should subscribe to the `detached` event of the newly created `amqp10` link object.]*/
+                  this._linkObject.on('detached', this._detachHandler);
+                  /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_007: [The `ReceiverLink` object should subscribe to the `errorReceived` event of the newly created `amqp10` link object.]*/
+                  this._linkObject.on('errorReceived', this._errorHandler);
+                  this._linkObject.on('message', this._messageHandler);
+                  /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_020: [The `attach` method shall call the `callback` if the link was successfully attached.]*/
+                  this._fsm.transition('attached', callback);
+                } else {
+                  debug('client forceDetached us already - cleaning up');
+                  amqp10link.forceDetach();
+                }
+                return null;
+              })
+              .catch((err) => {
+                debug('amqp10 failed to create receiver: ' + err.toString());
+                this._fsm.transition('detached', callback, err);
+              });
+          },
+          detach: (callback) => this._fsm.transition('detaching', callback),
+          forceDetach: (err) => this._fsm.transition('detached', undefined, err),
```

Comment:
>forceDetach: (err) => this._fsm.transition('detached', undefined, err), [](start = 10, length = 71)

Links could be present.  I think forceDetach should always _try_ to close and null out variables, otherwise this object will be in an indeterminate state.

Reply:
by "links could be present", I mean "the amqp10link object could be attached here.  Not likely, but still possible.

---
In reply to: [139543130](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139543130) [](ancestors = 139543130)

Reply:
I don't think it can, because 1/ the link object is nulled when entering detached 2/ before "saving" it in a private instance member, the callback checks that it's still in "attaching" state meaning if somehow forceDetach was called during the "attaching" state we would not have a link to forceDetach.

---
In reply to: [139543378](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139543378) [](ancestors = 139543378,139543130)

Reply:
OK, fair enough.  I was thinking paranoid, but I trust your opinion here.

---
In reply to: [139563612](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139563612) [](ancestors = 139563612,139543378,139543130)